tests/ndc/ndc.go

```diff
@@ -0,0 +1,2468 @@
+	historyClient := s.cluster.GetHistoryClient()
+
+	versions := []int64{3, 13, 2, 202, 302, 402, 602, 502, 802, 1002, 902, 702, 1102}
+	for _, version := range versions {
+		runID := uuid.New()
+		historySize := int64(0)
+
+		var historyBatch []*historypb.History
+		s.generator = test.InitializeHistoryEventGenerator(s.namespace, s.namespaceID, version)
+
+		for s.generator.HasNextVertex() {
+			events := s.generator.GetNextVertices()
+			historyEvents := &historypb.History{}
+			for _, event := range events {
+				historyEvents.Events = append(historyEvents.Events, event.GetData().(*historypb.HistoryEvent))
+			}
+			historySize += s.sizeOfHistoryEvents(historyEvents.Events)
+			historyBatch = append(historyBatch, historyEvents)
+		}
+
+		versionHistory := s.eventBatchesToVersionHistory(nil, historyBatch)
+		s.applyEvents(
+			workflowID,
+			runID,
+			workflowType,
+			taskqueue,
+			versionHistory,
+			historyBatch,
+			historyClient,
+		)
+		s.verifyEventHistorySize(workflowID, runID, historySize)
+		s.verifyEventHistory(workflowID, runID, historyBatch)
+	}
+}
+
+func (s *NDCFunctionalTestSuite) TestMultipleBranches() {
+
+	s.setupRemoteFrontendClients()
+	workflowID := "ndc-multiple-branches-test" + uuid.New()
+
+	workflowType := "event-generator-workflow-type"
+	taskqueue := "event-generator-taskQueue"
+
+	// cluster has initial version 1
+	historyClient := s.cluster.GetHistoryClient()
+
+	versions := []int64{102, 2, 202}
+	versionIncs := [][]int64{{1, 10}, {11, 10}}
+	versionInc := versionIncs[rand.Intn(len(versionIncs))]
+	for _, version := range versions {
+		runID := uuid.New()
+		historySize := int64(0)
+
+		var baseBranch []*historypb.History
+		baseGenerator := test.InitializeHistoryEventGenerator(s.namespace, s.namespaceID, version)
+		baseGenerator.SetVersion(version)
+
+		for i := 0; i < 10 && baseGenerator.HasNextVertex(); i++ {
+			events := baseGenerator.GetNextVertices()
+			historyEvents := &historypb.History{}
+			for _, event := range events {
+				historyEvents.Events = append(historyEvents.Events, event.GetData().(*historypb.HistoryEvent))
+			}
+			historySize += s.sizeOfHistoryEvents(historyEvents.Events)
+			baseBranch = append(baseBranch, historyEvents)
+		}
+		baseVersionHistory := s.eventBatchesToVersionHistory(nil, baseBranch)
+
+		var branch1 []*historypb.History
+		branchVersionHistory1 := versionhistory.CopyVersionHistory(baseVersionHistory)
+		branchGenerator1 := baseGenerator.DeepCopy()
+		for i := 0; i < 10 && branchGenerator1.HasNextVertex(); i++ {
+			events := branchGenerator1.GetNextVertices()
+			historyEvents := &historypb.History{}
+			for _, event := range events {
+				historyEvents.Events = append(historyEvents.Events, event.GetData().(*historypb.HistoryEvent))
+			}
+			historySize += s.sizeOfHistoryEvents(historyEvents.Events)
+			branch1 = append(branch1, historyEvents)
+		}
+		branchVersionHistory1 = s.eventBatchesToVersionHistory(branchVersionHistory1, branch1)
```

Comment:
It look like you are repeating this code number of times. May be we need a function?